backups/monthly_backup.sh

```diff
@@ -0,0 +1,216 @@
+    echo "$USAGE"
+    echo "Missing the required p4_backup_bucket parameter."
+    exit 1
+  fi
+
+  shift $((OPTIND-1))
+  P4_BACKUP_BUCKET=$1
+}
+
+#######################################
+# Validates the Perforce environment prior to running the backup.
+# Globals:
+#   None
+# Arguments:
+#   Perforce Server ID string
+#   Dry run flag
+#######################################
+validate_environment() {
+  local server_id=$1
+  local dry_run=$2
+
+  if [[ ! -n "${P4PORT}" ]]; then
+    err "P4PORT cannot be empty"
+    exit 1
+  fi
+
+  if [[ ! -n "${server_id}" ]]; then
+    err "Server ID cannot be empty"
+    exit 1
+  fi
+
+  if [[ $dry_run -eq 1 ]]; then
+    log "Executing in dry-run mode"
+  fi
+}
+
+cleanup () {
+  logs=$(find "${LOGS_ROOT}"/full_backup_log_* -type f -printf '%T@ %p\n' | sort -rn | cut -d" " -f 2 | tail -n +$((${MONTHLY_BACKUP_LOGS_TO_KEEP} + 1)))
+  for i in ${logs}; do
+    log "rm -rf ${i}"
+    rm -rf "$i"
+  done
+}
+
+main() {
+  # Load the profile to pick up p4 client settings
+  source "${HOME}/.profile"
+
+  # Parse command-line options
+  parse_options "$@"
+
+  mkdir -p "${LOGS_ROOT}"
+
+  CURRENT_LOG=${LOGS_ROOT}/full_backup_log_${TODAY}.txt
+
+  readonly CURRENT_LOG
+  readonly DRY_RUN
+  readonly IS_EDGE
+  readonly LOGS_ROOT
+  readonly P4_BACKUP_BUCKET
+
+  if [[ "${DRY_RUN}" -ne 1 ]]; then
+    usercheck
+  fi
+
+  local server_id
+  server_id=$(/usr/bin/p4 info | grep ServerID | awk '{print $2}')
+  if [[ ! -n "${server_id}" && "${DRY_RUN}" -eq 1 ]]; then
+    server_id=default
+  fi
+
+  local server_root
+  server_root=$(/usr/bin/p4 info | grep "Server root:"  | awk '{print $3}')
+  if [[ ! -d "${server_root}" ]]; then
+    err "Could not find the Helix Core server root"
+    exit 1
+  fi
+
+  local depot_root
+  depot_root=$(p4d -r "${server_root}" -cshow | grep "server.depot.root = " | awk '{print $4}')
+
+  local depot_path="${server_root}"/"${depot_root}"
+  if [[ ! -d "${depot_path}" ]]; then
+    err "Could not find the depot root"
+    exit 1
+  fi
+
+
+  log "Beginning backup for server ${server_id}..."
+
+  validate_environment "${server_id}" "${DRY_RUN}"
+
+  if [[ "${DRY_RUN}" -eq 1 ]]; then
+    log "Will execute perform_backup ${server_id} ${depot_path} ${DRY_RUN} ${IS_EDGE}"
+  else
+    perform_backup "${server_id}" "${depot_path}" "${DRY_RUN}" "${IS_EDGE}"
+    cleanup
+  fi
+
+  log "Backup of ${server_id} complete"
+}
+
+#######################################
+# Executes the backup by copying all required files to GCS.
+# Globals:
+#   TODAY
+# Arguments:
+#   Perforce Server ID string
+#   Dry run flag
+#   Edge flag
+#######################################
+perform_backup() {
+  set +e
+  local server_id=$1
+  local depot_path=$2
+  local dry_run=$3
+  local is_edge=$4
+
+  # Backing up versioned files
+  log "Backing up the versioned files at ${depot_path}"
+
+  if [[ "${is_edge}" -eq 0 ]]; then
+    gsutil \
+      -m cp -r "${depot_path}" 2>&1 \
```

Comment:
Why do you have multiple STDERR into STDOUT redirections inside a single command?

Reply:
Why, doesn't it redirect better if it has multiple ones? :) Fixed!